### PR TITLE
experimental: reactive pytest

### PR DESCRIPTION
**Pull request by dmadisetti**
## 📝 Summary

Follow up to: #3892

This allows eligible cells to automatically and reactively run pytest in a notebook:

image

It also hooks into the runtime, so tests don't have to run from scratch

This is currently hidden under an experimental flag for a few reasons:

1. I noticed live editing the cells sometimes puts them in an invalid state requiring another run (pytest loads the module off disk, and I think there's a bit of a lag)
2. similar to top level functions, I think that it seems like a bit of magic for when the functionality turns on. I think some more documentation and UX communication might be needed before an intentional release.
3. My unit tests aren't super robust and I haven't tested out every edge case
4. Would nice to get some feedback before pushing too much further
 
Just a note, this is tightly integrated with pytest, I don't think there are many other meaningful frameworks now nose is dead, but marimo should attempt to be agnostic.

## 📜 Reviewers

@akshayka OR @mscolnick @koaning

### CODE REVIEW

**Review by akshayka — COMMENTED** [review]
Quite interesting! I haven't really reviewed code yet, but I wanted to confirm that the idea is that even when moved out of experimental, this would be opt-in, correct? Otherwise I think it would be too magical

File: marimo/_runtime/runner/hooks_post_execution.py [hunk 1]
```diff
@@ -339,6 +340,26 @@ def format_output() -> formatting.FormattedOutput:
...
 
 
+@kernel_tracer.start_as_current_span("run_pytest")
+def _attempt_pytest(
+    cell: CellImpl,
+    runner: cell_runner.Runner,
+    run_result: cell_runner.RunResult,
+) -> None:
+    del run_result
+    if cell._test:
```
Comment (akshayka) [inline marimo/_runtime/runner/hooks_post_execution.py:350]: Only run if the runner isn't lazy?

Comment (dmadisetti) [reply marimo/_runtime/runner/hooks_post_execution.py:350]: Implemented as a conditional post-execution hook. With the experimental flag is in place, this would be a no-op with lazy

But definitely for removing the flag. I was thinking maybe a separate run button?

**Comment by mscolnick** [comment]
I'd make it opt-out, otherwise it won't be as discoverable. 

And otherwise, why else would you add pytest to your notebooks

**Comment by akshayka** [comment]
> I'd make it opt-out, otherwise it won't be as discoverable.
> 
> And otherwise, why else would you add pytest to your notebooks

You might just call a function `test_` something, without "adding pytest" to your notebook. That might seem odd, but with a large number of users that will happen. It would be jarring if just defining a function caused it to run. If someone wants to write unit tests in a notebook (a somewhat nice but compelling use case), they will be sufficiently motivated to find it in the docs.

**Review by akshayka — APPROVED** [review]
🔧

**Comment by akshayka** [comment]
@dmadisetti mind resolving the merge conflict? (sorry) otherwise good to go, thanks.
